packages/compass-connections/src/stores/store-context.tsx

```diff
@@ -0,0 +1,298 @@
+import React, { useCallback, useState } from 'react';
+import type {
+  MapStateToProps,
+  ReactReduxContextValue,
+  TypedUseSelectorHook,
+} from 'react-redux';
+import {
+  connect as reduxConnect,
+  createStoreHook,
+  createDispatchHook,
+  createSelectorHook,
+} from 'react-redux';
+import type {
+  configureStore,
+  ConnectionId,
+  ConnectionState,
+} from './connections-store-redux';
+import {
+  cancelEditConnection,
+  connect as connectionsConnect,
+  connectionsEventEmitter,
+  createNewConnection,
+  disconnect,
+  duplicateConnection,
+  editConnection,
+  getDataServiceForConnection,
+  removeAllRecentConnections,
+  removeConnection,
+  saveEditedConnectionInfo,
+  showNonGenuineMongoDBWarningModal,
+  toggleConnectionFavoritedStatus,
+  importConnections,
+} from './connections-store-redux';
+import type { Store } from 'redux';
+import {
+  getConnectionTitle,
+  type ConnectionInfo,
+} from '@mongodb-js/connection-info';
+import { createServiceLocator } from 'hadron-app-registry';
+import { isEqual } from 'lodash';
+
+type ConnectionsStore = ReturnType<typeof configureStore> extends Store<
```

Comment:
What I'm doing here is I'm trying to extract the exact types of the connection store from the `configureStore` function instead of just defining them manually. In theory I can replicate those closely by just writing them manually, but the benefit of inferring them like that is that the `dispatch` type in particular is taking into account the exact thunk arguments passed in this function, which means that if by accident we change those and they don't much the ThunkAction definition in the store code, I won't be able to dispatch those without TS complaining. This is helpful because the actions, useDispatch, and the store definition are completely separate from each other (because that's how it is in redux with the redux toolkit), so you can't really enforce this with types in any other way. Not sure if it makes sense or makes it even more confusing, happy to clarify more 🙂 

As for `infer` keyword itself, it's a neat feature of TS that allows you to unwrap / derive types from existing ones. For example if you'd want to unwrap a resolved type of a promise, it can look something like this:

```ts
type SomePromise = Promise<{ foo: number }>
type UnwrappedPromise = SomePromie extends Promise<infer Result> ? Result : never; // UnwrappedPromise is { foo: number }
```

My code is just this, but x3 times for all the arguments in the `Store` generic type. Doesn't need to be a generic either, the built-in typescript `Parameters` type is doing something similar to get function arguments out of any function type you pass to it:

```ts
type Parameters<MaybeFunction> = MaybeFunction extends (...args: infer Args) => any ? Args : never;
```